components/renku_data_services/storage/db.py

```diff
@@ -183,12 +182,15 @@ async def filter_projects_by_access_level(
         self, user: base_models.APIUser, project_ids: list[str], minimum_access_level: authz_models.Role
     ) -> list[str]:
         """Get a list of projects of which the user is a member with a specific access level."""
-        if not user.is_authenticated:
+        if not user.is_authenticated or not project_ids:
             return []
 
         scope = authz_models.Scope.WRITE if minimum_access_level == authz_models.Role.OWNER else authz_models.Scope.READ
+        user_project_ids = []
 
-        user_id = cast(str, user.id)
-        user_project_ids = await self.project_authz.get_user_projects(requested_by=user, user_id=user_id, scope=scope)
+        for project_id in project_ids:
+            has_permission = await self.project_authz.has_permission(user=user, project_id=project_id, scope=scope)
+            if has_permission:
+                user_project_ids.append(project_id)
```

Comment:
this still does one query per project. Instead I'd add a new method on project_authz that does an [in](https://docs.sqlalchemy.org/en/20/core/operators.html#in-comparisons) query for all of them at once.

Reply:
Thanks Ralf! We always call the `filter_projects_by_access_level` method with one project ID except in the [get_storage](https://github.com/SwissDataScienceCenter/renku-data-services/blob/1a6044abeb39cf7bf0f1d578bff7a25c38c3df24/components/renku_data_services/storage/db.py#L57) method. `get_storage` itself is always called with a single project ID although this can change in future. I've added a new method to `authz` as you've suggested in https://github.com/SwissDataScienceCenter/renku-data-services/pull/178/commits/9f1ddde2359b5637c362317f50ae1ffe6cf04c80.